### ISSUE TITLE
Wasted space when rendering trees

### ISSUE THREAD

**Issue by riedelcastro** [explosion/displacy#6]
displacy looks great, but I feel it could use much less space to present the same information. I used build whatswrongwith my nlp:

https://code.google.com/archive/p/whatswrong/
https://github.com/schmmd/whatswrong (github clone by someone else)

Nowhere as slick as displacy (and just a java app). But I felt the dependency graph rendering managed to display things quite compactly without loosing information:

![img](http://i1-win.softpedia-static.com/screenshots/What-s-Wrong-With-My-NLP_1.png) 

At least for my application, it would be great to have a bit more compactness, and a bigger chance to see the whole sentence without scrolling right.

**Comment by ines** [comment]
Yeah, I know what you mean – there are always trade-offs. displaCy was originally just an interactive demo, so the main aim was to make it easy to read and understand, even for people without an NLP background. So the current style felt a bit more natural.

But now that I think about it, it could be nice to have a "compact mode" for people who are mostly using the displacy.js library and are integrating it with their projects, Jupyter etc. 

It shouldn't actually be that much work either – at the moment, displaCy already calculates 4 coordinates per curve,  the start point, end point and the 2 cubic bezier points:

![](http://i.imgur.com/2219sMb.png)

So instead of drawing a curve using those coordinates, the "compact mode" could simply draw a path connecting those four points. Using a much smaller `distance`, the result would look pretty similar to the output What's Wrong With My NLP produces.

Will test this and report back!

**Comment by riedelcastro** [comment]
Yes, that could work! Also note that by using a little more straight line the labels don't need to be curved anymore. For some longer labels the curvature meant that they couldn't be read easily anymore (unless one uses a lot of space between tokens). 

Another thing I did in whatswrong was to calculate the height of the array based on the number of arrows it needs to cover. This meant the graph didn't need much vertical space either.

**Comment by ines** [comment]
Yeah, I was thinking about simply connecting these four points using a straight line and then decreasing the arrow height in general to make it even more compact.

I still remember the headache from trying to get the arrow height right... using `end - start` as the factor to calculate the height of the curves would cause gigantic arcs and too much wasted space for long dependencies, so displaCy now compiles a set of all different lengths that occur in the parse, and then uses the index of a dependency's length in that set as the factor (see [my blog post](https://explosion.ai/blog/displacy-js-nlp-visualizer#arcs) for reference, where I'm showing the before/after). This will probably work even better for the "compact" mode.

**Comment by andreasvc** [comment]
FWIW, I think the compactness is not so much about the arcs, but about word spacing. Currently words have fixed spacing, but short words require less space. This is the main reason why the one by whatswrong can fit longer sentences in the same space. You would have to compute the x position of each word by adding up the space each word requires.
